README.md
# MicroModel

A really basic ORM-like form and table mapper, for working with Silex, Symfony Forms, and Doctrine DBAL;
supports single tables without relationships.

***

## Example table

Here’s a table we’ll use for the rest of the code samples.

	`planets`

	| id       | name    | orbital_period | last_updated |
	| (PK, AI) | (text)  | (number)       | (text)       |
	------------------------------------------------------
	| 1        | Mercury | 87.97          | 1982-10-28   |
	| 2        | Venus   | 224.70         | 1980-05-21   |
	| 3        | Earth   | 365.25         | 1981-06-04   |

## How to use

1. Install with Composer.

	```js
	{
		"require": {
			"thomasjbradley/micromodel": "<2.0.0"
		}
	}
	```

2. Create a PHP class in your Silex application that extends MicroModel.
	**The class must be named identically to the table.**
	Capitalization doesn’t matter, the class name/table name will be converted to lowercase.

	```php
	<?php

	class Planets extends MicroModel
	{
		public function defineSchema ()
		{
			// Define all the table's fields
		}
	}
	```

3. Then make a new instance of your model, passing the Silex\Application.

	```php
	<?php
	$app = new Silex\Application();
	$planets = new Planets($app);
	$planetsList = $planets->all();
	```

## Defining fields

When defining fields,
the options array inherits everything from [Symfony\Form](http://symfony.com/doc/current/book/forms.html) options arrays.

**Always define the primary key field first.**

```php
<?php

use Symfony\Component\Validator\Constraints as Assert;

class Planets extends MicroModel
{
	public function defineSchema ()
	{
		// The primary key MUST always come first
		$this->defineField('id', 'integer');

		$this->defineField('name', 'text', array(
			'constraints' => array(new Assert\NotBlank())
			, 'set' => function ($val) {
				return filter_var($val, FILTER_SANITIZE_STRING);
			}
		));

		$this->defineField('orbital_period', 'number', array(
			'constraints' => array(new Assert\Number())
			, 'precision' => 2
			, 'set' => function ($val) {
				return filter_var($val, FILTER_SANITIZE_NUMBER_FLOAT, FILTER_FLAG_ALLOW_FRACTION);
			}
		));

		$this->defineField('last_updated', 'date', array(
			'constraints' => array(new Assert\NotBlank(), new Assert\Date())
			, 'format' => 'yyyy-MM-dd'
			, 'display' => false
			, 'set' => function ($val) {
				if (is_object($val)) return $val;
				return new DateTime($val);
			}
		));
	}
}
```

### Field definition extra options

MicroModel adds a few extra options to the array.

- `set` (function) — An optional setter function.
	Allows for data type conversion and sanitization.
- `display` (boolean) — Optional; whether the field should be shown in forms or not.

## Methods

### ☛ __construct( Silex\Application *$app* [, mixed *$clauses* = null ] )

Set up the model and optionally read a single item immediately by specifying `$clauses`.

- `$app` — the Silex\Application object.
- `$clauses` — passed directly to `read()`. See `read()` for more details.

```php
<?php
$planets = new Planets($app, 1);
echo $planets->name; // Mercury
```

### ☛ defineField( *$name* [, string *$type* = 'text' [, array *$options* = array() ]] )

Define a new field on the model to match a field in the table. *Usually called from within the `defineSchema()` method.* [Refer to defining fields](#defining-fields).

- `$name` — The name of the field, spelled identically to the table’s field name.
- `$type` — One of the [Symfony\Form](http://symfony.com/doc/current/book/forms.html) field types. They match very well to the [Doctrine\DBAL](http://www.doctrine-project.org/projects/dbal.html) field types.
- `$options` — Any of the Symfony\Form field options. Refer to [field registration extra options](#field-registration-extra-options) for more options.

**@return** — `$this`

### ☛ find( [ string|array *$order* = null [, array *$where* = array() ]] )

Get a bunch of results from the table, optionally sorting them.
Without any arguments `find()` will return every entry in the database.
*Will return an array of the model objects.*

- `$order` — the field names & direction for the order clause.
- `$where` — arrays of WHERE clause conditions, following this format: `array('field', 'comparison', 'value')`.

**@return** — an array of objects, each object is an instance of your model.

```php
<?php
$planets = new Planets($app);
$planetsList = $planets->find();

foreach ($planetsList as $planet) {
	echo $planet->name;
}

// Since each item in the array is your model object, you could do this
$planetsList[0]->name = 'Neptune';
$planetsList[0]->update();

// Using the $order argument
$planets->find('name ASC');
$planets->find(array('name ASC', 'orbital_period DESC'));

// Using the $where argument
$planets->find(null, array(
	array('orbital_period', '>', 200)
	, array('name', 'LIKE', '%e%')
));

// Using $order and $where
$planets->find('name ASC', array(
	array('orbital_period', '>', 200)
));
```

### ☛ create()

`INSERT` the current object into the table using the property values.
After insertion, the primary key field is populated with `lastInsertId`.
**The data is not validated before creation is attempted.**

**@return** — `$this`

```php
<?php
$planets = new Planets($app);
$planets->name = 'Jupiter';
$planets->orbital_period = 4332.59;
$planets->last_updated = new DateTime();
$planets->create();
echo $planets->id; // 4
```

### ☛ read( mixed *$clauses* )

Read a single entry from the table, converting all the fields to properties of the object.

- `$clauses` — conditions for reading a single entry in the table.
	- `int|string` — the value for the individual item’s primary key.
	- `array` — arrays of WHERE clause conditions, following this format: `array('field', 'comparison', 'value')`.

**@return** — `$this`

```php
<?php
$planets = new Planets($app);

// Use the primary key to select an item
// Equates to WHERE id = 1
$planet = $planets->read(1);
echo $planet->name; // Mercury

// Set up a WHERE clause with arrays
// Equates to WHERE name = 'Earth'
$planet = $planets->read(array(
	array('name', '=', 'Earth')
));
echo $planet->name; // Earth
```

### ☛ update()

`UPDATE` the current object in the table using the property values.
Uses the field marked as primary key for the `WHERE` clause.
**The data is not validated before updating is attempted.**

**@return** — `$this`

```php
<?php
$planets = new Planets($app, 2);
$planets->last_updated = new DateTime();
$planets->update();
```

### ☛ delete()

`DELETE` the current object from the table.
Uses the field marked as primary key for the `WHERE` clause.

**@return** — `$this`

```php
<?php
$planets = new Planets($app, 3);
$planets->delete();
```

### ☛ getForm( [ boolean *$csrfProtection* = true ] )

Returns a [Symfony\Form](http://symfony.com/doc/current/book/forms.html) object for the model.
All constraints and options from the field registration are used.

- `$csrfProtection` — flag for enabling/disabling CSRF proection; helpful primarily for APIs

**@return** — Symfony\Form

```php
<?php
$planets = new Planets($app, 1);
$form = $planets->getForm();
// $form->bind($request);
// $form->isValid();
// $form->createView();

// When writing a JSON API
$form = $planets->getForm(false);
// Symfony's form bind() method expects an array, so force json_decode to use associative arrays
$form->bind(json_decode($request->getContent(), true));

if ($form->isValid()) {
	$movie->create();
	$app->abort(204);
}
```

### ☛ jsonSerialize()

Double duty: returns a simplified version of all the fields’ values in an associative array,
and in PHP/5.4 it is the JsonSerializer implementation.

**@return** — an `array` containing all the fields and their values

```php
<?php
$planets = new Planets($app, 2);

// PHP/5.4
return $app->json($planets);

// PHP/5.3
return $app->json($planets->jsonSerialize());
// Because PHP/5.3 doesn't have the JsonSerializer interface
// Will work equally as well in PHP/5.4
```

### ☛ isValid()

Validates the information in the object against the field constraints.

**@return** — `boolean`

```php
<?php
$planets = new Planets($app);
$planets->name = 'Saturn';
$planets->orbital_period = 10759.22;
$planets->last_updated = new DateTime();
$planets->isValid(); // true
```

### ☛ getErrors()

Return an array of all the validation error messages produced by Symfony\Form.
Generally called after `isValid()` if you need to see the error messages.
If there are no error messages the array is empty.

**@return** — `array`

***

## License

MicroModel is licensed under the [BSD 3-Clause license](https://github.com/thomasjbradley/micromodel/blob/master/BSD-3-CLAUSE-LICENSE.txt).


CHANGELOG.md
# Changelog

**1.1.3 (Aug. 9, 2012)**

- Fixed typo in offsetGet, contributed by @tyaga

**1.1.2 (Jul. 31, 2012)**

- Added an exception in defineSchema() in case the method is not recreated in the model

**1.1.1 (Jul. 30, 2012)**

- Renamed a few methods to make MicroModel be more consistent with Mongoose

**1.1.0 (Jul. 28, 2012)**

- Added JSON serialization of the model
- Added WHERE clause support to `read()` and `all()`
- Changed the `isValid()` method to return only true/false. Added `getErrors()` method
- Added CSRF protection flag to the `getForm()` method
- Bug fix: the iterator implementation for current was not returning the value
- Bug fix: determining the table name from the class name did not account for namespaces
- Bug fix: fields were being set twice because of PDO::FETCH_BOTH
- Bug fix: choice fields weren't supported

**1.0.1 (Jul. 27, 2012)**

- Bug fix: `create()` wasn't setting the primary key
- Bug fix: `register()` wasn't returning the model object
- Internal consistency updates

**1.0.0 (Jul. 26, 2012)**

- Initial Release


lib/MicroModel.php
<?php
/**
 * MicroModel: A really basic ORM-like form and table mapper, for working with Silex, Symfony\Forms, and Doctrine\DBAL
 * Supports single tables without relationships
 * @author Thomas J Bradley <hey@thomasjbradley.ca>
 * @link http://github.com/thomasjbradley/micromodel
 * @copyright 2012 Thomas J Bradley
 * @license BSD-3-Clause <https://github.com/thomasjbradley/micromodel/blob/master/BSD-3-CLAUSE-LICENSE.txt>
 */

// Still support PHP/5.3, will be useless in PHP\5.4
if (!interface_exists('JsonSerializable')) {
  interface JsonSerializable {
    function jsonSerialize ();
  }
}

abstract class MicroModel implements \ArrayAccess, \Iterator, \JsonSerializable {
  /**
   * Instance of Silex\Application
   * @var Silex\Application
   */
  protected $__app;

  /**
   * Instance of the Doctrine DBAL Connection
   * @var Doctrine\DBAL\Connection
   */
  protected $__db;

  /**
   * Holds all the table's fields, values, and form constraints
   * @var array
   */
  protected $__fields = array();

  /**
   * Sets up DB connection; registers table fields; optionally reads single row
   * @param Silex\Application $app The Silex application with Doctrine DBAL
   * @param mixed $clauses Passed directly to read(); {@see MicroModel::read()}
   * @throws Exception
   */
  public function __construct (\Silex\Application $app, $clauses = null) {
    $this->__app = $app;

    if (!isset($app['db']) || get_class($app['db']) != 'Doctrine\DBAL\Connection') {
      throw new Exception('The Silex application must have the DoctrineServiceProvider registered.');
    }

    if (!isset($app['validator']) || get_class($app['validator']) != 'Symfony\Component\Validator\Validator') {
      throw new Exception('The Silex application must have the ValidatorServiceProvider registered.');
    }

    if (!isset($app['form.factory']) || get_class($app['form.factory']) != 'Symfony\Component\Form\FormFactory') {
      throw new Exception('The Silex application must have the FormServiceProvider registered.');
    }

    $this->__db = $this->__app['db'];

    /**
     * @deprecated registerFields() deprecated since version 1.1.1, to be removed in 2.0.0
     *             See {@link defineSchema()}
     */
    if (method_exists($this, 'registerFields')) {
      $this->registerFields();
    } else {
      $this->defineSchema();
    }

    if (!is_null($clauses))
      $this->read($clauses);
  }

  /**
   * Generic getter for all the fields
   * @param string $field The field's name
   * @return mixed
   */
  public function __get ($field) {
    return $this->__fields[$field]['data'];
  }

  /**
   * Generic setter for all the field
   * @param string $field The field's name
   * @param mixed $val The new value for the field
   */
  public function __set ($field, $val) {
    if (isset($this->__fields[$field]['set'])) {
      $this->__fields[$field]['data'] = $this->__fields[$field]['set']($val);
    } else {
      $this->__fields[$field]['data'] = $val;
    }
  }

  /**
   * ArrayAccess implementation for offsetSet: changes the field's value
   * @param string $offset The field's name
   * @param array $value A replacement value for the field
   * @return void
   */
  public function offsetSet ($offset, $value) {
    $this->__set($offset, $value);
  }

  /**
   * ArrayAccess implementation for offsetExists
   * @param string $offset The field's name
   * @return bool
   */
  public function offsetExists ($offset) {
    return isset($this->__fields[$offset]);
  }

  /**
   * ArrayAccess implementation for offsetUnset: empties the field's value
   * @param string $offset The field's name
   * @return void
   */
  public function offsetUnset ($offset) {
    $this->__fields[$offset]['data'] = null;
  }

  /**
   * ArrayAccess implementation for offsetGet: returns the field's value
   * @param string $offset The field's name
   * @return mixed
   */
  public function offsetGet ($offset) {
    return $this->__fields[$offset]['data'];
  }

  /**
   * Iterator implementation of rewind
   * @return void
   */
  public function rewind () {
    prev($this->__fields);
  }

  /**
   * Iterator implementation of current
   * @return array All the field's details
   */
  public function current () {
    $opts = current($this->__fields);

    return $opts['data'];
  }

  /**
   * Iterator implementation of key
   * @return string The field's name
   */
  public function key () {
    return key($this->__fields);
  }

  /**
   * Iterator implementation of next
   * @return void
   */
  public function next () {
    next($this->__fields);
  }

  /**
   * Iterator implementation of valid
   * @return bool
   */
  public function valid () {
    return isset($this->__fields[key($this->__fields)]);
  }

  /**
   * Registers a table's field as a field
   * @param string $field The field's name
   * @param string $type The field's type, matches Symfony\Form types
   * @param array $options The field's options, matches Symfony\Form options
   * @return $this
   */
  public function defineField ($field, $type = 'text', $options = array()) {
    if (!isset($options['type']))
      $options['type'] = $type;

    if (!isset($options['data']))
      $options['data'] = null;

    $this->__fields[$field] = $options;

    return $this;
  }

  /**
   * Abstract function that must be implemented by the model
   * Used to register all the table field types
   * @todo Make abstract method in 2.0.0
   * @throws Exception
   * @return void
   */
  public function defineSchema () {
    throw new Exception(sprintf(
      'The defineSchema() method is missing in your %s MicroModel.'
      , get_class($this)
    ));
  }

  /**
   * Returns all, or some, of the items from the table for this type
   * @param array $order The ORDER BY clause, e.g. name ASC
   * @param array $where An array of arrays comprising the WHERE clause
   * @return array Each item mapped to the model class
   */
  public function find ($order = null, $where = array()) {
    $sql = sprintf('SELECT * FROM %s', $this->getTableName());
    $whereClauses = array();

    if (!empty($where)) {
      foreach ($where as $clause) {
        $whereClauses[] = sprintf('%s %s :%s', $clause[0], $clause[1], $clause[0]);
      }

      $sql .= sprintf(' WHERE %s', implode(' AND ', $whereClauses));
    }

    if (!empty($order)) {
      if (is_string($order))
        $order = array($order);

      $sql .= sprintf(' ORDER BY %s', implode(',', $order));
    }

    $stmt = $this->__db->prepare($sql);

    if (!empty($where)) {
      foreach ($where as $clause) {
        $stmt->bindValue($clause[0], $clause[2]);
      }
    }

    $stmt->execute();
    $results = $stmt->fetchAll(PDO::FETCH_ASSOC);
    $items = array();

    // Hack because PDO::FETCH_CLASS doesn't work reliably
    foreach ($results as $item) {
      $class = get_class($this);
      $obj = new $class($this->__app);

      foreach ($item as $k => $v) {
        $obj->__set($k, $v);
      }

      $items[] = $obj;
    }

    return $items;
  }

  /**
   * Creates a new entry in the table
   * @return $this
   */
  public function create () {
    reset($this->__fields);
    $pk = key($this->__fields);
    $form = $this->getForm();
    $placeholders = array();

    foreach ($form as $k => $v) {
      $placeholders[$k] = ':' . $k;
    }

    $sql = sprintf(
      'INSERT INTO %s (%s) VALUES (%s)'
      , $this->getTableName()
      , implode(',', array_keys($placeholders))
      , implode(',', $placeholders)
    );

    $stmt = $this->__db->prepare($sql);

    foreach ($form as $k => $v) {
      if (is_object($v->getData()) && get_class($v->getData()) == 'DateTime') {
        $stmt->bindValue($k, $v->getData(), $this->__fields[$k]['type']);
      } else {
        $stmt->bindValue($k, $v->getData());
      }
    }

    $stmt->execute();
    $this->$pk = $this->__db->lastInsertId();

    return $this;
  }

  /**
   * Gets a single item from the table
   * @param int|string|array $clauses Int|String: the value for the primary key; Array: arrays comprising the WHERE clause
   * @return $this
   */
  public function read ($clauses) {
    $whereClauses = array();

    if (!is_array($clauses)) {
      reset($this->__fields);
      $pk = key($this->__fields);
      $clauses = array(array($pk, '=', $clauses));
    }

    foreach ($clauses as $clause) {
      $whereClauses[] = sprintf('%s %s :%s', $clause[0], $clause[1], $clause[0]);
    }

    $sql = sprintf(
      'SELECT * FROM %s WHERE %s'
      , $this->getTableName()
      , implode(' AND ', $whereClauses)
    );

    $stmt = $this->__db->prepare($sql);

    foreach ($clauses as $clause) {
      $stmt->bindValue($clause[0], $clause[2]);
    }

    $stmt->execute();
    $results = $stmt->fetch(PDO::FETCH_ASSOC);

    foreach ($this->__fields as $k => $v) {
      $this->$k = $results[$k];
    }

    return $this;
  }

  /**
   * Updates this entry in the table
   * @return $this
   */
  public function update () {
    reset($this->__fields);
    $pk = key($this->__fields);
    $form = $this->getForm();
    $updates = array();

    foreach ($form as $k => $v) {
      $updates[] = sprintf('%s = :%s', $k, $k);
    }

    $sql = sprintf(
      'UPDATE %s SET %s WHERE %s = :%s'
      , $this->getTableName()
      , implode(',', $updates)
      , $pk
      , $pk
    );

    $stmt = $this->__db->prepare($sql);
    $stmt->bindValue($pk, $this->__get($pk));

    foreach ($form as $k => $v) {
      if (is_object($v->getData()) && get_class($v->getData()) == 'DateTime') {
        $stmt->bindValue($k, $v->getData(), $this->__fields[$k]['type']);
      } else {
        $stmt->bindValue($k, $v->getData());
      }
    }

    $stmt->execute();

    return $this;
  }

  /**
   * Deletes this entry in the table
   * @return $this
   */
  public function delete () {
    reset($this->__fields);
    $pk = key($this->__fields);

    $this->__db->delete(
      $this->getTableName()
      , array($pk => $this->__get($pk))
    );

    return $this;
  }

  /**
   * Gets a Symfony\Form object based on registered fields
   * @param boolean $csrfProtection Whether to use CSRF protection or not
   * @return Symfony\Component\Form\Form
   */
  public function getForm ($csrfProtection = true) {
    $builder = $this->__app['form.factory']->createBuilder('form', $this, array(
      'csrf_protection' => $csrfProtection
    ));

    foreach (array_slice($this->__fields, 1, null, true) as $k => $v) {
      if (isset($v['display']) && $v['display'] == false)
        continue;

      $options = $v;
      unset($options['type']);
      unset($options['set']);
      unset($options['display']);

      $builder->add($k, $v['type'], $options);
    }

    $form = $builder->getForm();

    return $form;
  }

  /**
   * Confirms that the current model object is valid or not
   * @return boolean
   */
  public function isValid () {
    $validator = $this->__app['validator'];

    foreach ($this->getValues(true) as $k => $v) {
      if (!isset($this->__fields[$k]['constraints']))
        continue;

      foreach ($this->__fields[$k]['constraints'] as $constraint) {
        $possibleError = $validator->validateValue($v, $constraint);

        if (count($possibleError) != 0)
          return false;
      }
    }

    return true;
  }

  /**
   * Returns a list of all the validation error messages
   * @return array Error list if invalid; empty array otherwise
   */
  public function getErrors () {
    $errors = array();
    $validator = $this->__app['validator'];

    foreach ($this->getValues(true) as $k => $v) {
      if (!isset($this->__fields[$k]['constraints']))
        continue;

      foreach ($this->__fields[$k]['constraints'] as $constraint) {
        $possibleError = $validator->validateValue($v, $constraint);

        if (count($possibleError) != 0) {
          $errors[$k] = array();
          $errors[$k][get_class($constraint)] = $possibleError;
        }
      }
    }

    return $errors;
  }

  /**
   * Converts this item to a JSON object
   * In PHP/5.3 this method must be called directly before json_encode()
   *   json_encode($myModel->jsonSerialize())
   * @return array The simplified fields and values
   */
  public function jsonSerialize () {
    reset($this->__fields);
    $pk = key($this->__fields);
    $data = array(
      $pk => $this->__get($pk)
    );

    $form = $this->getForm();

    foreach ($form as $k => $v) {
      if (is_object($v->getData()) && get_class($v->getData()) == 'DateTime') {
        $data[$k] = $v->getViewData();
      } else {
        $data[$k] = $v->getData();
      }
    }

    return $data;
  }

  /**
   * Converts the class's name into the table name
   * @return string
   */
  protected function getTableName () {
    $class = explode('\\', get_class($this));

    return strtolower(end($class));
  }

  /**
   * Gets all the values for this item and puts them in an array
   * @param boolean $withPk Whether the primary key item should be included or not
   * @return array
   */
  protected function getValues ($withPk = true) {
    if (!$withPk) {
      $values = array_slice($this->__fields, 1, null, true);
    } else {
      $values = $this->__fields;
    }

    foreach ($values as $k => $v) {
      $values[$k] = $this->__get($k);
    }

    return $values;
  }

  /**
   * @deprecated Deprecated since version 1.1.1, to be removed in 2.0.0
   *             Use {@link defineField()} instead
   */
  public function register ($field, $type = 'text', $options = array()) {
    $this->defineField($field, $type, $options);
  }

  /**
   * @deprecated Deprecated since version 1.1.1, to be removed in 2.0.0
   *             Use {@link find()} instead
   */
  public function all ($order = null, $where = array()) {
    return $this->find($order, $where);
  }
}
